ui/pages/Login/reducers.js

```diff
@@ -19,6 +20,42 @@ export const login = (values) => {
   }
 }
 
+export const googleLogin = () => {
+  return new HttpRequestHelper('/api/login_google',
+    (responseJson) => {
+      // Redirect to google auth website
+      const width = 600
+      const height = 800
+      const left = (window.screen.width - width) / 2
+      const top = (window.screen.height - height) / 2
+      const params = `scrollbars=no,status=no,location=no,toolbar=no,menubar=no,
+        width=${width},height=${height},left=${left},top=${top}`
+      const win = window.open(responseJson.data, 'Google Sign In', params)
+      win.focus()
+    },
+    (e) => {
+      throw new SubmissionError({ _error: [e.message] })
+    },
+  ).get()
+}
+
+export const loadGoogleAuthResult = (location) => {
+  return (dispatch) => {
+    dispatch({ type: REQUEST_GOOGLE_AUTH_RESULT })
+    new HttpRequestHelper('/api/login_oauth2callback',
+      (responseJson) => {
+        dispatch({ type: RECEIVE_DATA, updatesById: responseJson })
```

Comment:
RECEIVE_DATA triggers a TON of state reloads, its means that one of the core data state fields has been updated. If you aren't expecting one of the main data fields to be updated, you should dispatch a different event just for login. What exactly is in the `responseJson`?

Reply:
Okay. The purpose of `RECEIVE_DATA` here is just to set the `loading` to false. `responseJson` doesn't have anything useful.